### PR TITLE
feat: add parallelism for dsl.ParallelFor. Fixes #4089

### LINKED ISSUE #4089: [Feature Request] Add parallelism for dsl.ParallelFor
**Use Case**

Large data mining job is often split into many small jobs. Given the limit shared resource of external services (e.g., DataFlow), we can only run a few small jobs simultaneously. 

Global parallelism works to some degree but lack flexibility, e.g., with global parallelism at 1, any in-cluster task can block launching jobs to external services.

Argo supports template-level parallelism
https://github.com/argoproj/argo/blob/master/examples/parallelism-nested-workflow.yaml#L19
https://github.com/argoproj/argo/blob/master/examples/parallelism-nested-dag.yaml#L15

**Feature request**

```python
with dsl.ParallelFor(loopidy_doop, parallelism=2) as item:
    // DAG in loop here
```
There will be at most 2 loop-DAG running in parallel.

### PR DESCRIPTION
**Pull request by NikeNano**
**Description:**
This PR updates the ParallelFor in order to allow for setting the allowed parallelism. Setting parallelism will set the parallelism on the sub-Dag level similar to this example: https://github.com/argoproj/argo/blob/master/examples/parallelism-nested-dag.yaml. Related to: :https://github.com/kubeflow/pipelines/issues/4089

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

   PR titles examples:
    * `fix(frontend): fixes empty page. Fixes #1234`
       Use `fix` to indicate that this PR fixes a bug.
    * `feat(backend): configurable service account. Fixes #1234, fixes #1235`
       Use `feat` to indicate that this PR adds a new feature. 
    * `chore: set up changelog generation tools`
       Use `chore` to indicate that this PR makes some changes that users don't need to know.
    * `test: fix CI failure. Part of #1234`
        Use `part of` to indicate that a PR is working on an issue, but shouldn't close the issue when merged.

- [ ] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
    * **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.
    *  After this PR is merged, create a cherry-pick PR to add these changes to the release branch. (For more information about creating a cherry-pick PR, see the [Kubeflow Pipelines release guide](https://github.com/kubeflow/pipelines/blob/master/RELEASE.md#option--git-cherry-pick).)

### CODE REVIEW

File: sdk/python/kfp/compiler/compiler.py [hunk 1]
```diff
@@ -587,6 +589,7 @@ def _create_dag_templates(self, pipeline, op_transformers=None, op_to_templates_
     #   op_name_to_parent_groups also contains the recursive opsgroups
     #   condition_params from _get_condition_params_for_ops also contains the recursive opsgroups
     #   groups does not include the recursive opsgroups
+    # HERE HERE HERE
```
Comment (NikeNano) [inline sdk/python/kfp/compiler/compiler.py:592]:
```suggestion

```

File: sdk/python/kfp/compiler/compiler.py [hunk 2]
```diff
@@ -611,6 +613,8 @@ def _create_dag_templates(self, pipeline, op_transformers=None, op_to_templates_
 
     templates = []
     for opsgroup in opsgroups.keys():
+      # HERE WE SEEM TO MAKE IT TO A GROUP. 
+      # HOW FAR CAN WE GO...
```
Comment (NikeNano) [inline sdk/python/kfp/compiler/compiler.py:617]:
```suggestion
```

File: samples/core/loop_parallelism/loop_parallelism.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            command=["sh", "-c"],
+            arguments=["echo op2 %s" % item.B_b],
+        )
+
+    op_out = dsl.ContainerOp(
+        name="my-out-cop",
+        image="library/bash:4.4.23",
+        command=["sh", "-c"],
+        arguments=["echo %s" % my_pipe_param],
+    )
```
Comment (Ark-kun) [inline samples/core/loop_parallelism/loop_parallelism.py:42]:
```suggestion
@kfp.components.create_component_from_func
def print_op(s: str):
    print(s)

@dsl.pipeline(name='my-pipeline')
def pipeline2(my_pipe_param=10):
    loop_args = [{'A_a': 1, 'B_b': 2}, {'A_a': 10, 'B_b': 20}]
    with dsl.ParallelFor(loop_args, parallelism=1) as item:
        print_op(item)
        print_op(item.A_a)
        print_op(item.B_b)
```
```

**Comment by Ark-kun** [comment]
Thank you for this work.

**Comment by NikeNano** [comment]
> I've proposed to simplify the sample

Sure, I think your suggestion sounds good @Ark-kun

**Comment by NikeNano** [comment]
Seems like the CLA is not signed for the email which the suggestion is linked to @Ark-kun

**Comment by Bobgy** [comment]
Yes, suggestion feature is so far broken with suggestion. So unfortunately, I'd suggest not using it.

I can force approve this PR for you.

**Comment by NikeNano** [comment]
> Yes, suggestion feature is so far broken with CLA. So unfortunately, I'd suggest not using it.
> 
> I can force approve this PR for you.

ok, :( Thanks!

**Comment by hlu09** [comment]
@NikeNano ,  this implementation seems only limit the parallelism (1 in the sample) to one instance of 
(print_op(item),  print_op(item.A_a),   print_op(item.B_b)). If ParallelFor generates 10 such instances (sub-DAG),  10 first print_ops will run at the same time.

**Comment by NikeNano** [comment]
The current examples has a bug, will make a PR for it. Example workflow: 

``` 
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: my-pipeline-
  annotations: {pipelines.kubeflow.org/kfp_sdk_version: 0.5.1, pipelines.kubeflow.org/pipeline_compilation_time: '2020-07-09T11:12:21.683680',
    pipelines.kubeflow.org/pipeline_spec: '{"name": "my-pipeline"}'}
  labels: {pipelines.kubeflow.org/kfp_sdk_version: 0.5.1}
spec:
  entrypoint: my-pipeline
  templates:
  - name: for-loop-for-loop-5b8096fe-1
    parallelism: 10
    inputs:
      parameters:
      - {name: loop-item-param-5b8096fe}
      - {name: loop-item-param-5b8096fe-subvar-A_a}
      - {name: loop-item-param-5b8096fe-subvar-B_b}
    dag:
      tasks:
      - name: print-op
        template: print-op
        arguments:
          parameters:
          - {name: loop-item-param-5b8096fe, value: '{{inputs.parameters.loop-item-param-5b8096fe}}'}
      - name: print-op-2
        template: print-op-2
        arguments:
          parameters:
          - {name: loop-item-param-5b8096fe-subvar-A_a, value: '{{inputs.parameters.loop-item-param-5b8096fe-subvar-A_a}}'}
      - name: print-op-3
        template: print-op-3
        arguments:
          parameters:
          - {name: loop-item-param-5b8096fe-subvar-B_b, value: '{{inputs.parameters.loop-item-param-5b8096fe-subvar-B_b}}'}
  - name: my-pipeline
    dag:
      tasks:
      - name: for-loop-for-loop-5b8096fe-1
        template: for-loop-for-loop-5b8096fe-1
        arguments:
          parameters:
          - {name: loop-item-param-5b8096fe, value: '{{item}}'}
          - {name: loop-item-param-5b8096fe-subvar-A_a, value: '{{item.A_a}}'}
          - {name: loop-item-param-5b8096fe-subvar-B_b, value: '{{item.B_b}}'}
        withItems:
        - {A_a: 1, B_b: 2}
        - {A_a: 10, B_b: 20}
  - name: print-op
    container:
      args: [--s, '{{inputs.parameters.loop-item-param-5b8096fe}}']
      command:
      - python3
      - -u
      - -c
      - |
        def print_op(s):
            print(s)

        import argparse
        _parser = argparse.ArgumentParser(prog='Print op', description='')
        _parser.add_argument("--s", dest="s", type=str, required=True, default=argparse.SUPPRESS)
        _parsed_args = vars(_parser.parse_args())

        _outputs = print_op(**_parsed_args)
      image: python:3.7
    inputs:
      parameters:
      - {name: loop-item-param-5b8096fe}
    metadata:
      annotations: {pipelines.kubeflow.org/component_spec: '{"implementation": {"container":
          {"args": ["--s", {"inputValue": "s"}], "command": ["python3", "-u", "-c",
          "def print_op(s):\n    print(s)\n\nimport argparse\n_parser = argparse.ArgumentParser(prog=''Print
          op'', description='''')\n_parser.add_argument(\"--s\", dest=\"s\", type=str,
          required=True, default=argparse.SUPPRESS)\n_parsed_args = vars(_parser.parse_args())\n\n_outputs
          = print_op(**_parsed_args)\n"], "image": "python:3.7"}}, "inputs": [{"name":
          "s", "type": "String"}], "name": "Print op"}', pipelines.kubeflow.org/component_ref: '{}'}
  - name: print-op-2
    container:
      args: [--s, '{{inputs.parameters.loop-item-param-5b8096fe-subvar-A_a}}']
      command:
      - python3
      - -u
      - -c
      - |
        def print_op(s):
            print(s)

        import argparse
        _parser = argparse.ArgumentParser(prog='Print op', description='')
        _parser.add_argument("--s", dest="s", type=str, required=True, default=argparse.SUPPRESS)
        _parsed_args = vars(_parser.parse_args())

        _outputs = print_op(**_parsed_args)
      image: python:3.7
    inputs:
      parameters:
      - {name: loop-item-param-5b8096fe-subvar-A_a}
    metadata:
      annotations: {pipelines.kubeflow.org/component_spec: '{"implementation": {"container":
          {"args": ["--s", {"inputValue": "s"}], "command": ["python3", "-u", "-c",
          "def print_op(s):\n    print(s)\n\nimport argparse\n_parser = argparse.ArgumentParser(prog=''Print
          op'', description='''')\n_parser.add_argument(\"--s\", dest=\"s\", type=str,
          required=True, default=argparse.SUPPRESS)\n_parsed_args = vars(_parser.parse_args())\n\n_outputs
          = print_op(**_parsed_args)\n"], "image": "python:3.7"}}, "inputs": [{"name":
          "s", "type": "String"}], "name": "Print op"}', pipelines.kubeflow.org/component_ref: '{}'}
  - name: print-op-3
    container:
      args: [--s, '{{inputs.parameters.loop-item-param-5b8096fe-subvar-B_b}}']
      command:
      - python3
      - -u
      - -c
      - |
        def print_op(s):
            print(s)

        import argparse
        _parser = argparse.ArgumentParser(prog='Print op', description='')
        _parser.add_argument("--s", dest="s", type=str, required=True, default=argparse.SUPPRESS)
        _parsed_args = vars(_parser.parse_args())

        _outputs = print_op(**_parsed_args)
      image: python:3.7
    inputs:
      parameters:
      - {name: loop-item-param-5b8096fe-subvar-B_b}
    metadata:
      annotations: {pipelines.kubeflow.org/component_spec: '{"implementation": {"container":
          {"args": ["--s", {"inputValue": "s"}], "command": ["python3", "-u", "-c",
          "def print_op(s):\n    print(s)\n\nimport argparse\n_parser = argparse.ArgumentParser(prog=''Print
          op'', description='''')\n_parser.add_argument(\"--s\", dest=\"s\", type=str,
          required=True, default=argparse.SUPPRESS)\n_parsed_args = vars(_parser.parse_args())\n\n_outputs
          = print_op(**_parsed_args)\n"], "image": "python:3.7"}}, "inputs": [{"name":
          "s", "type": "String"}], "name": "Print op"}', pipelines.kubeflow.org/component_ref: '{}'}
  arguments:
    parameters: []
  serviceAccountName: pipeline-runner
```
This workflow should be have similar to [this](https://github.com/argoproj/argo/blob/master/examples/parallelism-nested-dag.yaml) in the argo docs. Which as I understand has the parallelism in the sub-DAG:s, from the argo example: 

```
# Example on specifying parallelism on the outer DAG and limiting the number of its
# children DAGs to be run at the same time.
# 
# As the parallelism of A is 2, only two of the three DAGs (b2, b3, b4) will start 
# running after b1 is finished, and the left DAG will run after either one is finished.

```

Maybe I am missing something, could you point me in the right direction @hlu09 ? Thanks!

**Comment by NikeNano** [comment]
Ohhh now I get it @hlu09 I think you are correct! I will check this further tonight.

**Comment by hlu09** [comment]
@NikeNano  One possible solution is to add a new "with Parallelism(2):" op-group, which limits the parallelism of its **immediate** child DAGs or tasks. Thus, if a ParallelFor under it generates 10 sub-DAGs at run-time, only 2 will be executed at a time.  

https://github.com/kubeflow/pipelines/compare/master...hlu09:parallel_limit

```python
with Parallelism(2):
   with dsl.ParallelFor(loop_args) as item:
```

while more generic (it can be used with any set of ops), it is not as nice as having `with dsl.ParallelFor(loop_args, parallelism=2)`

**Pull request merged**